contribs/drt/src/main/java/org/matsim/contrib/drt/prebooking/PrebookingManager.java

```diff
@@ -180,30 +196,204 @@ public PassengerRequest retrieveRequest(MobsimAgent agent, Leg leg) {
 	// Housekeeping of requests
 
 	private IdMap<Request, RequestItem> requests = new IdMap<>(Request.class);
+	private IdSet<Request> unscheduleUponVehicleAssignment = new IdSet<>(Request.class);
 
 	private class RequestItem {
-		// this class looks minimal for now, but will be extended with canceling
-		// functionality
 		final PassengerRequest request;
 
+		Id<DvrpVehicle> vehicleId = null;
+		boolean onboard = false;
+
 		RequestItem(PassengerRequest request) {
 			this.request = request;
 		}
 	}
 
+	void notifyPickup(double now, AcceptedDrtRequest request) {
+		RequestItem item = requests.get(request.getId());
+
+		if (item != null) {
+			// may be null, we treat all (also non-prebooked) requests here
+			item.onboard = true;
+		}
+	}
+
+	void notifyDropoff(Id<Request> requestId) {
+		requests.remove(requestId);
+	}
+
+	@Override
+	public void handleEvent(PassengerRequestScheduledEvent event) {
+		RequestItem item = requests.get(event.getRequestId());
+
+		if (item != null) {
+			item.vehicleId = event.getVehicleId();
+		}
+
+		if (unscheduleUponVehicleAssignment.contains(event.getRequestId())) {
+			// this is the case if a request has been rejected / canceled after submission
+			// but before scheduling
+			unscheduler.unscheduleRequest(event.getTime(), event.getVehicleId(), event.getRequestId());
+			unscheduleUponVehicleAssignment.remove(event.getRequestId());
+		}
```

Comment:
True, I will investigate this further. There might be an issue then that in the next time step the agent enters before it can be unscheduled. I'll check how to do this cleanly.